File: src/main/java/com/huawei/wallet/nsp/NetworkService.java
/*
 * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.huawei.wallet.nsp;

import com.huawei.wallet.util.ConfigUtil;

import com.alibaba.fastjson.JSONObject;

import org.apache.logging.log4j.util.Strings;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestTemplate;

/**
 * Implementation of invoking gateway service.
 *
 * @since 2019-07-26
 */
public class NetworkService {
    private static final RestTemplate REST_TEMPLATE = new RestTemplate();

    public static String getToken(String clientId, String clientSecret) {
        HttpHeaders header = new HttpHeaders();
        header.setContentType(MediaType.APPLICATION_FORM_URLENCODED);

        MultiValueMap<String, String> map = new LinkedMultiValueMap<>();
        map.add("grant_type", "client_credentials");
        map.add("client_id", clientId);
        map.add("client_secret", clientSecret);

        String tokenUrl = ConfigUtil.instants().getValue("gw.tokenUrl");

        // Send the http request and get response.
        HttpEntity<MultiValueMap<String, String>> entity = new HttpEntity<>(map, header);
        ResponseEntity<JSONObject> exchange =
            REST_TEMPLATE.exchange(tokenUrl, HttpMethod.POST, entity, JSONObject.class);

        JSONObject response = exchange.getBody();

        // Return the token.
        if (response == null) {
            throw new NullPointerException("Get null token response.");
        }
        String accessToken = response.getString("access_token");
        if (Strings.isEmpty(accessToken)) {
            throw new NullPointerException("Get null access token.");
        }
        return accessToken;
    }
}

File: main/nsp/NetworkService.php
<?php
/*
 * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

/**
 * Implementation of invoking gateway service.
 *
 * @since 2020-05-25
 */
class NetworkService
{

    /**
     * Get access token with your App ID and App Secret.
     *
     * @param string $clientId your App ID on AGC website.
     * @param string $clientSecret your App Secret on AGC website.
     * @return string the token string
     */
    public static function getToken(string $clientId, string $clientSecret)
    {
        $config = parse_ini_file('../config/config.ini');
        $request = curl_init();
        curl_setopt($request, CURLOPT_SSL_VERIFYPEER, 0);
        curl_setopt($request, CURLOPT_URL, $config['gw.tokenUrl']);
        curl_setopt($request, CURLOPT_POST, 1);
        curl_setopt($request, CURLOPT_HTTPHEADER, array('Content-Type: application/x-www-form-urlencoded'));
        curl_setopt($request, CURLOPT_POSTFIELDS,
            "grant_type=client_credentials&client_secret=" . $clientSecret . "&client_id=" . $clientId);
        curl_setopt($request, CURLOPT_RETURNTRANSFER, true);
        $response = curl_exec($request);

        if ($response == null) {
            throw new BadMethodCallException("Get token failed.\nError code: " . curl_errno($request) .
                ". Error message: " . curl_error($request) . "\n");
        }

        $returnCode = curl_getinfo($request, CURLINFO_HTTP_CODE);
        curl_close($request);
        if ($returnCode != "200") {
            throw new UnexpectedValueException("Get token failed.\nError code: " . $returnCode .
                ". Error message: " . $response . "\n");
        }

        return json_decode($response, true)["access_token"];
    }
}